File: examples/destructuring.rb
Obj = Struct.new(:key, :value) do
  def read
    [key, value]
  end
end

obj = Obj.new('foo', 'bar')
key, value = obj.key, obj.value
puts "#{key} #{value}"
key, value = obj.read
puts "#{key} #{value}"

File: examples/destructuring.c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct { const char *key; const char *value; } Obj;
static void read_obj(const Obj *obj, const char **key, const char **value) { *key = obj->key; *value = obj->value; }
int main(void) { Obj obj = {"foo", "bar"}; const char *key = obj.key, *value = obj.value; printf("%s %s\n", key, value); read_obj(&obj, &key, &value); printf("%s %s\n", key, value); return 0; }
